internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -170,6 +176,40 @@ func (r *AgentServiceConfigReconciler) Reconcile(origCtx context.Context, req ct
 		return reconcile.Result{}, nil
 	}
 
+	if instance.DeletionTimestamp.IsZero() {
+		if !funk.ContainsString(instance.GetFinalizers(), agentServiceConfigFinalizerName) {
+			controllerutil.AddFinalizer(instance, agentServiceConfigFinalizerName)
+			if err := r.Update(ctx, instance); err != nil {
+				log.WithError(err).Error("failed to add finalizer to AgentServiceConfig")
+				return ctrl.Result{Requeue: true}, err
+			}
+		}
+	} else {
+		// do cleanup and remove finalizer
+		statefulSet := &appsv1.StatefulSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      imageServiceName,
+				Namespace: r.Namespace,
+			},
+		}
+		if err := r.Get(ctx, client.ObjectKeyFromObject(statefulSet), statefulSet); err != nil && !errors.IsNotFound(err) {
+			log.WithError(err).Error("failed to get image service stateful set for cleanup")
+			return ctrl.Result{Requeue: true}, err
+		}
+		if err := r.cleanupImageServiceFinalizer(ctx, statefulSet); err != nil {
+			log.WithError(err).Error("failed to cleanup image service stateful set")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		controllerutil.RemoveFinalizer(instance, agentServiceConfigFinalizerName)
```

Comment:
It may be wise to wait until the StatefulSet is actually gone. Consider someone who wants to delete and then re-create an ACI, perhaps even with a script. As soon as the ACI is deleted, they make a new one, and some error occurs because the reconciler can't create a new stateful set since the old one is still being deleted. Given the parent/child tree relationship, the safe thing to do seems to be to not delete parents until children are completely cleaned up.

Reply:
Won't that eventually resolve though?
Alternatively if we've already removed the finalizer on the statefulset couldn't you make the same argument for any of the deployments?

Reply:
It'll resolve, but would need to be handled gracefully by the controller.

I'm not following your point about deployments?

Reply:
>would need to be handled gracefully by the controller.

You mean the statefulset controller here, right?

> I'm not following your point about deployments?

I mean, I don't think there's anything special about a statefulset in the situation you described.
There's a parent/child relationship between ASC and every object we create so if someone deletes it and recreates a new one we'll have issues with every object we create until the old one is removed. This applies to deployments, secrets, services, etc. I don't see why we would treat statefulsets specially here outside of ensuring the finalizer is removed.

Reply:
Fair point. Since we're certain by this point that the finalizer on the StatefulSet has been removed, I'm fine leaving it as-is.